drafts/swift-async-await-cycle.md

```diff
@@ -0,0 +1,122 @@
+---
+title: The retain cycle of Swift async/await
+subtitle: How to inadvertently extend object lifetime
+domain: software-engineering-corner.hashnode.dev
+tags: swift, swift-concurrency, ios, macos, watchos, tvos, visionos, apple, vapor, bluetooth, concurrency, asynchronous, async, await, async-await, developer, learning, software-development, programming-tips, software-engineering, code, coding
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1724686396858/4k_l7bLPj.jpg?auto=format
+publishAs: simonseyer
+hideFromHashnodeCommunity: false
+saveAsDraft: true
+--- 
+In a recent project, we observed one of those bugs that lets every developer jubilate. It was only happening rarely and seemed to manifest in varying ways. 
+
+But first some context: We were running a Swift/Vapor server which controls a generic Zühlke Bluetooth emulation dongle (connected via USB) in order to emulate medical devices. This talks to an iOS app via Bluetooth (running on a real phone) that is under test by a custom (medical) XCTest-based test framework. In summary, quite some moving parts.
+
+![Real project setup](https://cdn.hashnode.com/res/hashnode/image/upload/v1724685658047/IBP4odcIg.png?auto=format)
+
+When executing tests in sequence (i.e. not individually), they would sometimes fail because of a communication failure. Curiously, most messages were still transmitted successfully. After some digging, we realised that on failure, the tests that executed *before* had one thing in common: they terminated the connection forcefully. This was fine by itself but led us to think that somehow the tests were not executed in isolation.
```

Comment:
Perhaps clarify what kind of communication failure it is to ensure the context is clear?
When executing tests in sequence (i.e. not individually), they would sometimes fail because of a **Bluetooth** communication failure. Curiously, most messages were still transmitted successfully.